common/src/main/java/org/astraea/common/balancer/BalancerConsoleImpl.java

```diff
@@ -0,0 +1,446 @@
+
+  private final AtomicBoolean stopped = new AtomicBoolean(false);
+
+  public BalancerConsoleImpl(Admin admin, Function<Integer, Optional<Integer>> jmxPortMapper) {
+    this.admin = admin;
+    this.jmxPortMapper = jmxPortMapper;
+  }
+
+  @Override
+  public Set<String> tasks() {
+    return Collections.unmodifiableSet(taskPhases.keySet());
+  }
+
+  @Override
+  public Optional<TaskPhase> taskPhase(String taskId) {
+    return Optional.ofNullable(taskPhases.get(taskId));
+  }
+
+  @Override
+  public Generation launchRebalancePlanGeneration() {
+    return new Generation() {
+      private String taskId;
+      private Balancer balancer;
+      private AlgorithmConfig algorithmConfig;
+      private Duration timeout = Duration.ofSeconds(1);
+      private boolean checkNoOngoingMigration = true;
+
+      @Override
+      public Generation setTaskId(String taskId) {
+        this.taskId = taskId;
+        return this;
+      }
+
+      @Override
+      public Generation setBalancer(Balancer balancer) {
+        this.balancer = balancer;
+        return this;
+      }
+
+      @Override
+      public Generation setGenerationTimeout(Duration timeout) {
+        this.timeout = timeout;
+        return this;
+      }
+
+      @Override
+      public Generation setAlgorithmConfig(AlgorithmConfig config) {
+        this.algorithmConfig = config;
+        return this;
+      }
+
+      @Override
+      public Generation checkNoOngoingMigration(boolean enable) {
+        this.checkNoOngoingMigration = enable;
+        return this;
+      }
+
+      @Override
+      public CompletionStage<Balancer.Plan> generate() {
+        final var taskId = Objects.requireNonNull(this.taskId);
+        final var balancer = Objects.requireNonNull(this.balancer);
+        final var timeout = Objects.requireNonNull(this.timeout);
+        final var config = Objects.requireNonNull(this.algorithmConfig);
+        final var sensors =
+            Stream.concat(
+                    config.clusterCostFunction().metricSensor().stream(),
+                    config.moveCostFunction().metricSensor().stream())
+                .collect(Collectors.toUnmodifiableList());
+        if (this.checkNoOngoingMigration) BalancerConsoleImpl.this.checkNoOngoingMigration();
+        final var sourceCluster =
+            admin.topicNames(false).thenCompose(admin::clusterInfo).toCompletableFuture().join();
+        synchronized (this) {
+          checkNotClosed();
+          return tasks
+              .compute(
+                  taskId,
+                  (id, previousTask) -> {
+                    if (previousTask != null)
+                      throw new IllegalStateException("Conflict task ID: " + taskId);
+                    taskPhases.put(taskId, TaskPhase.Searching);
+                    return new BalanceTaskImpl(
+                        taskId,
+                        sourceCluster,
```

Comment:
> sourceCluster和checkNoOngoingMigration

目前的邏輯是

1. 收集叢集狀態和檢查條件
2. 接受計劃，調度執行

改成你說的手法會變成

1. 接受計劃，調度執行
2. 收集叢集狀態和檢查條件

如果 "收集叢集狀態和檢查條件" 這個條件一定會噴錯（比如目前正好有 Ongoing migration），那麼前面的手法可以比較快讓使用者知道有錯誤發生，而後者（你提到的方法），一定會讓計劃被調度，而調度之後會立刻失敗。我覺得目前前者的做法比較有 fail fast 的效果。在 `POST /balancer` 那邊也能比較及時的像使用者反映問題。改成後者的方法的話，使用者可能 `POST /balancer` 調度之後，等 30 秒讓演算法計算完，`GET /balancer` 時才發現其實計劃最早一開始的時候就失敗了。

然後我不覺得改成放到非同步裡面有什麼顯著的好處，反而會失去 fail fast 的優勢。

Reply:
> 我覺得目前前者的做法比較有 fail fast 的效果

這是一個不錯的議題

> 我覺得目前前者的做法比較有 fail fast 的效果

所以你把非同步的程式碼硬轉成同步是為了 fail fast 嗎？你講的情境是“使用者”的需要吧？如果這兩者都成立的話，為何用非同步的包裝就無法讓使用者做 "fail fast" 呢？使用者可以拿到非同步的包裝後，自行等待一段時間確認狀態轉移就好，這樣的做法會比現在直接 hard-code 寫在impl裡面只為了你提到的“某種使用情境”來得更有彈性

> 然後我不覺得改成放到非同步裡面有什麼顯著的好處，反而會失去 fail fast 的優勢。

這個命題可能也不太對，非同步和 fail fast 並不是互斥，這就是我前面留言提到，對於非同步模型不熟的話，很容易寫起來綁手綁腳（就像你覺得做非同步就無法 fail fast）

Reply:
> 所以你把非同步的程式碼硬轉成同步是為了 fail fast 嗎？

目前版本可以更快發生錯誤，不需要刻意 join 就能看到錯誤，且計劃不會被接受( `tasks` map 裡面不會有這個 taskId)

> 自行等待一段時間確認狀態轉移就好

這邊的一個問題是要等待的時間長度，或是要在 Searching 前面再添加一個狀態表達正在做檢查。

> 這個命題可能也不太對，非同步和 fail fast 並不是互斥

我所想的 fail 比你的 fail 更快，為了做到你想要的非同步效果，這和我這個速度的 fail 是互斥的。

我會試著改成你說的樣子，改了之後的好處大概就這個函數的平行程度會更好一點，但代價是程式碼實作的複雜度

Reply:
> 我所想的 fail 比你的 fail 更快，為了做到你想要的非同步效果，這和我這個速度的 fail 是互斥的。

可能有我沒想到的部分，以`checkNoOngoingMigration`為例，目前的方式是在產生 completablefuture後等待該結束並根據結果返回，而我提到的方法基本一樣，主要差別是別中斷completablefuture，繼續包裝成適合的返回物件。所以我不太確定為何會有誰快誰不快的狀況？

Reply:
> 改了之後的好處大概就這個函數的平行程度會更好一點，但代價是程式碼實作的複雜度

複雜度通常只會在包裹completablefuture時需要用特定的函式去串接，裡面其實一樣是寫 blocking code，如果會寫得異常複雜通常是有些 call chaining 沒有接好導致需要額外的轉換邏輯，這個就屬於實作面的調整

Reply:
> 目前版本可以更快發生錯誤，不需要刻意 join 就能看到錯誤

這句我也沒有看懂，現在程式碼裡面充滿著各種 join，`checkNoOngoingMigration`, `checkPlanConsistency`等等都有 join 強迫中斷的用法，請問你是指哪部分不需要刻意 join 就能看到錯誤?

Reply:
```java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.astraea.common.balancer;

import java.time.Duration;
import java.util.Collections;
import java.util.Comparator;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import org.astraea.common.admin.Admin;
import org.astraea.common.admin.ClusterBean;
import org.astraea.common.admin.Replica;
import org.astraea.common.balancer.algorithms.AlgorithmConfig;
import org.astraea.common.balancer.executor.RebalancePlanExecutor;

public class BalancerConsoleImpl implements BalancerConsole {

  private final Admin admin;

  private final Supplier<ClusterBean> clusterBeanSupplier;

  private final Map<String, CompletionStage<Balancer.Plan>> plans = new ConcurrentHashMap<>();

  private final Map<String, CompletionStage<Void>> executions = new ConcurrentHashMap<>();

  private final AtomicBoolean stopped = new AtomicBoolean(false);

  public BalancerConsoleImpl(Admin admin, Supplier<ClusterBean> clusterBeanSupplier) {
    this.admin = admin;
    this.clusterBeanSupplier = clusterBeanSupplier;
  }

  @Override
  public Set<String> tasks() {
    return Collections.unmodifiableSet(plans.keySet());
  }

  @Override
  public Optional<TaskPhase> taskPhase(String taskId) {
    var p = plans.get(taskId);
    if (p == null) return Optional.empty();
    if (p.toCompletableFuture().isCompletedExceptionally())
      return Optional.of(TaskPhase.Failed);
    if (!p.toCompletableFuture().isDone()) return Optional.of(TaskPhase.Searching);
    var e = executions.get(taskId);
    if (e == null) return Optional.of(TaskPhase.Searched);
    if (p.toCompletableFuture().isCompletedExceptionally())
      return Optional.of(TaskPhase.Failed);
    if (!p.toCompletableFuture().isDone()) return Optional.of(TaskPhase.Executing);
    return Optional.of(TaskPhase.Executed);
  }

  @Override
  public Generation launchRebalancePlanGeneration() {
    return new Generation() {
      private String taskId;
      private Balancer balancer;
      private AlgorithmConfig algorithmConfig;
      private Duration timeout = Duration.ofSeconds(1);
      private boolean checkNoOngoingMigration = true;

      @Override
      public Generation setTaskId(String taskId) {
        this.taskId = taskId;
        return this;
      }

      @Override
      public Generation setBalancer(Balancer balancer) {
        this.balancer = balancer;
        return this;
      }

      @Override
      public Generation setGenerationTimeout(Duration timeout) {
        this.timeout = timeout;
        return this;
      }

      @Override
      public Generation setAlgorithmConfig(AlgorithmConfig config) {
        this.algorithmConfig = config;
        return this;
      }

      @Override
      public Generation checkNoOngoingMigration(boolean enable) {
        this.checkNoOngoingMigration = enable;
        return this;
      }

      @Override
      public CompletionStage<Balancer.Plan> generate() {
        checkNotClosed();
        final var taskId = Objects.requireNonNull(this.taskId);
        final var balancer = Objects.requireNonNull(this.balancer);
        final var timeout = Objects.requireNonNull(this.timeout);
        final var config = Objects.requireNonNull(this.algorithmConfig);
        var checks =
            this.checkNoOngoingMigration
                ? BalancerConsoleImpl.this.checkNoOngoingMigration()
                : CompletableFuture.completedFuture(null);
        return plans.compute(
            taskId,
            (id, previousTask) -> {
              if (previousTask != null)
                throw new IllegalStateException("Conflict task ID: " + taskId);
              return checks
                  .thenCompose(ignored -> admin.topicNames(false).thenCompose(admin::clusterInfo))
                  .thenApply(
                      clusterInfo ->
                          balancer.offer(
                              clusterInfo,
                              clusterBeanSupplier.get(),
                              timeout,
                              AlgorithmConfig.builder(config)
                                  .metricSource(clusterBeanSupplier)
                                  .build()));
            });
      }
    };
  }

  @Override
  public Execution launchRebalancePlanExecution() {
    return new Execution() {
      private RebalancePlanExecutor executor;
      private Duration timeout;
      private boolean checkPlanConsistency = true;
      private boolean checkNoOngoingMigration = true;

      @Override
      public Execution setExecutor(RebalancePlanExecutor executor) {
        this.executor = executor;
        return this;
      }

      @Override
      public Execution setExecutionTimeout(Duration timeout) {
        this.timeout = timeout;
        return this;
      }

      @Override
      public Execution checkPlanConsistency(boolean enable) {
        this.checkPlanConsistency = enable;
        return this;
      }

      @Override
      public Execution checkNoOngoingMigration(boolean enable) {
        this.checkNoOngoingMigration = enable;
        return this;
      }

      @Override
      public CompletionStage<Void> execute(String taskId) {
        checkNotClosed();
        var planFuture = plans.get(taskId);
        if (planFuture == null)
          throw new IllegalArgumentException("No such balance task: " + taskId);
        var checks =
            this.checkNoOngoingMigration
                ? BalancerConsoleImpl.this.checkNoOngoingMigration()
                : CompletableFuture.completedFuture(null);

        return checks
            .thenCompose(ignored -> planFuture)
            .thenCompose(
                plan ->
                    this.checkPlanConsistency
                        ? BalancerConsoleImpl.this.checkPlanConsistency(plan)
                        : CompletableFuture.completedFuture(plan))
            .thenCompose(
                plan ->
                    executions.compute(
                        taskId,
                        (id, previousTask) -> {
                          if (previousTask != null)
                            throw new IllegalStateException("Conflict task ID: " + taskId);
                          return executor.run(admin, plan.solution.proposal(), timeout);
                        }));
      }
    };
  }

  @Override
  public void close() {
    if (stopped.get()) return;
    synchronized (this) {
      // reject further request
      stopped.set(true);
    }
  }

  private void checkNotClosed() {
    if (stopped.get()) throw new IllegalStateException("This BalanceConsole is already stopped");
  }

  private CompletionStage<Void> checkNoOngoingMigration() {
    return admin
        .topicNames(false)
        .thenCompose(admin::clusterInfo)
        .thenAccept(
            clusterInfo -> {
              var ongoingMigration =
                  clusterInfo
                      .replicaStream()
                      .filter(r -> r.isAdding() || r.isRemoving() || r.isFuture())
                      .map(Replica::topicPartitionReplica)
                      .collect(Collectors.toUnmodifiableSet());
              if (!ongoingMigration.isEmpty())
                throw new IllegalStateException(
                    "Another rebalance task might be working on. "
                        + "The following topic/partition has ongoing migration: "
                        + ongoingMigration);
            });
  }

  private CompletionStage<Balancer.Plan> checkPlanConsistency(Balancer.Plan plan) {
    final var before =
        plan
            .initialClusterInfo
            .replicaStream()
            .collect(Collectors.groupingBy(Replica::topicPartition))
            .entrySet()
            .stream()
            .collect(
                Collectors.toUnmodifiableMap(
                    Map.Entry::getKey,
                    e ->
                        e.getValue().stream()
                            // have to compare by isLeader instead of isPreferredLeader. since the
                            // leadership is what affects the direction of traffic load. Any
                            // bandwidth related cost function should calculate load by leadership
                            // instead of preferred leadership.
                            .sorted(
                                Comparator.comparing(Replica::isPreferredLeader)
                                    .reversed()
                                    .thenComparing(x -> x.nodeInfo().id()))
                            .map(x -> Map.entry(x.nodeInfo().id(), x.path()))
                            .collect(Collectors.toUnmodifiableList())));

    return admin
        .clusterInfo(plan.initialClusterInfo.topicNames())
        .thenApply(
            clusterInfo ->
                clusterInfo
                    .replicaStream()
                    .collect(Collectors.groupingBy(Replica::topicPartition))
                    .entrySet()
                    .stream()
                    .collect(
                        Collectors.toUnmodifiableMap(
                            Map.Entry::getKey,
                            e ->
                                e.getValue().stream()
                                    .sorted(
                                        Comparator.comparing(Replica::isPreferredLeader)
                                            .reversed()
                                            .thenComparing(x -> x.nodeInfo().id()))
                                    .map(x -> Map.entry(x.nodeInfo().id(), x.path()))
                                    .collect(Collectors.toUnmodifiableList()))))
        .thenApply(
            now -> {
              var mismatchPartitions =
                  before.entrySet().stream()
                      .filter(
                          e -> {
                            final var tp = e.getKey();
                            final var beforeList = e.getValue();
                            final var nowList = now.get(tp);
                            return !Objects.equals(beforeList, nowList);
                          })
                      .map(Map.Entry::getKey)
                      .collect(Collectors.toUnmodifiableSet());

              if (!mismatchPartitions.isEmpty())
                throw new IllegalStateException(
                    "The cluster state has been changed significantly. "
                        + "The following topic/partitions have different replica list(lookup the moment of plan generation): "
                        + mismatchPartitions);
              return plan;
            });
  }
}

```

我寫了一個範本給你參考一下，我們可以就這個為出發點討論可能我沒注意到/或是可以優化的地方

另外我注意到`AlgorithmConfig`依然帶有`Supplier<ClusterBean>`，我記得好像有討論過要移除？

Reply:
> 所以我不太確定為何會有誰快誰不快的狀況？

```java
var task =
    balancerConsole
        .launchRebalancePlanGeneration()
        .setTaskId(taskId)
        .setBalancer(balancer)
        .setAlgorithmConfig(request.algorithmConfig)
        .setGenerationTimeout(request.executionTime)
        .checkNoOngoingMigration(true)
        .generate();  // 1
task.toCompletableFuture().join() // 2
```

我是偏好 1 的時候噴錯，但你比較希望 2 的時候噴錯，我是說這邊的快慢

>> 目前版本可以更快發生錯誤，不需要刻意 join 就能看到錯誤
>>
> 請問你是指哪部分不需要刻意 join 就能看到錯誤?

這個 join 是上面 2 的 join

> 另外我注意到AlgorithmConfig依然帶有Supplier<ClusterBean>，我記得好像有討論過要移除？

這個部分在 https://github.com/skiptests/astraea/pull/1493#discussion_r1103635073 有提到

`Balancer#retryOffer` 需要新鮮的 `ClusterBean`，目前只能透過 `AlgorithmConfig` 拿到，這個部分的責任會等到這隻 PR 完成後，才會把這些工作放到 BalancerConsole 裡面執行，到時候 `AlgorithmConfig` 裡面的 `Supplier<ClusterBean>` 就能移除了

---

```java
  @Override
  public void close() {
    if (stopped.get()) return;
    synchronized (this) {
      // reject further request
      stopped.set(true);
    }
  }
```

剩下執行到一半的工作不需要終止嗎

Reply:
> 我是偏好 1 的時候噴錯，但你比較希望 2 的時候噴錯，我是說這邊的快慢

你說的 1 只是在背後偷偷呼叫的`join`，這樣應該不算是快慢的問題，差別只是包裹著例外是由誰戳破而已。對於使用者而言，他想直接戳破也行、想包裹著繼續 call chain 也行，2的好處就是非同步的彈性可以擴展給使用者

> 到時候 AlgorithmConfig 裡面的 Supplier<ClusterBean> 就能移除了

這邊應該還有一個題目是改成 `Balancer#offer(AlgorithmConfig)`，這樣之後balancer介面就可以固定下來，唯一會變的是`AlgorithmConfig`，這部分看要不要先改，以免之後越改越多，看你有沒有空，沒空的話我先把這些細節調整好

> 剩下執行到一半的工作不需要終止嗎

我貼的是範本，如果你打算放棄的話，我才會接手把剩下的全部做完